lib/shopify_api/bulk/query.ex

```diff
@@ -0,0 +1,172 @@
+defmodule ShopifyAPI.Bulk.Query do
+  alias ShopifyAPI.AuthToken
+  alias ShopifyAPI.Bulk.Cancel
+
+  @type status_response :: map()
+
+  @polling_query """
+  {
+    currentBulkOperation {
+      id
+      status
+      errorCode
+      createdAt
+      completedAt
+      objectCount
+      fileSize
+      url
+      partialDataUrl
+    }
+  }
+  """
+
+  @spec cancel(AuthToken.t(), String.t()) :: {:ok | :error, any()}
+  def cancel(token, bulk_query_id) do
+    query = """
+    mutation {
+      bulkOperationCancel(id: "#{bulk_query_id}") {
+        bulkOperation {
+          status
+        }
+        userErrors {
+          field
+          message
+        }
+      }
+    }
+    """
+
+    # {
+    #   "data": {
+    #     "bulkOperationCancel": {
+    #       "bulkOperation": {
+    #         "status": "CANCELING"
+    #       },
+    #       "userErrors": []
+    #     }
+    #   },
+    #   "extensions": {
+    #     "cost": {
+    #       "requestedQueryCost": 10,
+    #       "actualQueryCost": 10,
+    #       "throttleStatus": {
+    #         "maximumAvailable": 1000.0,
+    #         "currentlyAvailable": 990,
+    #         "restoreRate": 50.0
+    #       }
+    #     }
+    #   }
+    # }
+
+    case ShopifyAPI.graphql_request(token, query, 1) do
+      {:ok, %{response: %{"bulkOperationCancel" => resp}}} -> {:ok, resp}
+      error -> error
+    end
+  end
+
+  @spec exec(AuthToken.t(), String.t(), list()) :: {:ok, String.t()} | {:error, any()}
+  def exec(%AuthToken{} = token, query, opts) do
+    with bulk_query <- bulk_query_string(query),
+         {:ok, resp} <- ShopifyAPI.graphql_request(token, bulk_query, 10),
+         :ok <- handle_errors(resp),
+         bulk_query_id <- get_in(resp.response, ["bulkOperationRunQuery", "bulkOperation", "id"]),
+         {:ok, url} <- poll(token, bulk_query_id, opts[:polling_rate], opts[:max_poll_count]) do
+      {:ok, url}
+    else
+      {:error, :timeout, bulk_id} ->
+        Cancel.pretty_please(opts[:auto_cancel], token, bulk_id)
+
+      error ->
+        error
+    end
+  end
+
+  # Shopify returns a single newline which gets stripped and we are left with garbage,
+  # handle it nicely here.
+  def fetch(nil), do: {:ok, ""}
+
+  def fetch({:error, _} = error), do: error
+  def fetch({:ok, url}), do: fetch(url)
+
+  def fetch(url) when is_binary(url) do
+    url
+    |> HTTPoison.get()
+    |> case do
+      {:ok, %{body: jsonl}} -> {:ok, jsonl}
+      error -> error
+    end
+  end
+
+  def parse_response!(""), do: []
+  def parse_response!({:ok, jsonl}), do: parse_response!(jsonl)
+  def parse_response!({:error, msg}), do: raise(ShopifyAPI.Bulk.QueryError, msg)
```

Comment:
We need to add another case here, for when msg is a struct we raise `inspect(msg)`